## Code written by Michael Vella
## Generates figures S4 and S5 from:
## Development and testing of a novel Killer-Rescue self-limiting 
## gene drive system in Drosophila melanogaster
## Authors: Sophia H. Webster, Michael R. Vella, and Maxwell J. Scott

## Must have KR_functions.R, KR_data.csv, and offprobs.rds in working directory
library(dplyr)
library(reshape2)
library(ggplot2)
library(ggthemes)

source('KR_functions.R')

# read in data and manipulate to give genotype fractions
KRdat <- (read.csv("KR_data.csv",header=TRUE))
KRdat <- KRdat[order(KRdat$replicate),]
KRdat$experiment <- factor(KRdat$experiment)
KRdat[,4:10] <- KRdat[,4:10]/KRdat$adults 
KRdat <- KRdat %>% subset(select=-c(adults))
# read in offspring probability array:
# Parray[m,n,i] gives prob of offspring i from mating of female m and male n
# generated by a script but can be calculated by hand
offprobs <- readRDS('offprobs')
Parray <- offprobs[[1]]
genotype_names <- offprobs[[3]]
genotypes <- offprobs[[2]]        

###### Figure S4
## correlation between fitness parameters for the Gal80 rescue only experiment 
exp <- "G80"
dat <- KRdat %>% subset(experiment == exp)

## set input list for simulations and parameter estimation
input <- set_init(exp,genotype_names)
input[["Parray"]] <- Parray
# each row is a different combination of the two Gal80-only fitnesses
pgrid <- expand.grid(kkRR = as.list(seq(0.7,1,.01)),kkRr = as.list(seq(0.7,1,.01)))
# make a surface by producing the error sum of squares at each parameter combination
mysurf <- seq(1,nrow(pgrid)) %>% 
  vapply(function(x) KR_val_func(transto(unlist(pgrid[x,])),input,dat),c(1))

pgrid$ESS <- mysurf
pgrid$kkRR <- as.double(pgrid$kkRR)
pgrid$kkRr <- as.double(pgrid$kkRr)

# best fit
pgrid[which(pgrid$ESS==min(pgrid$ESS)),]

pgrid %>% ggplot(aes(x=kkRR,y=kkRr)) +
  geom_tile(aes(fill=log10(ESS))) +
  geom_contour(aes(z=log10(ESS))) +
  coord_equal() +
  scale_fill_distiller(direction=-1,
                       breaks=c(0.25,0,-0.25,-0.5),
                       labels=c("1.28","1.0","0.78","0.61"),
                       name="ESS",
                       guide=guide_colorbar(title.position = "top",
                                            frame.colour = "black",
                                            barheight = 10))+
  scale_x_continuous(name="kkRR fitness",expand = c(0, 0)) +
  scale_y_continuous(name="kkRr fitness",expand = c(0, 0)) +
  my_theme() #in KR_functions.R


###### Figure S5
## long-term simulations for DSCP-Gal4 killer;Gal80 rescue
exp <- "DG4"

# set up input with "best fit" parameters
input <- set_init(exp,genotype_names)
input[["Parray"]] <- Parray
input$w[input$pindex] <- c(0.70,0.73,0.83,0.77) #paramfit 

# release frequencies
rel_sizes <- c(2/3,2/3,2/3,1/3)
# number of generations to release
relnums <- c(1,2,5,5)

# run simulation for each release size/number
simdat_all <- lapply(seq_len(4),
                    function(x) {
                      # browser()
  rel_size <- rel_sizes[x]
  input$init <- c(rel_size,0,0,0,1-rel_size,0,0,0,0)
  input$tendin <- 2
  # simulate first generation after release
  simdat <- simfunc(input)
  rel <- 1
  # for each additional release, run another simulation starting from the previous last generation
  while (rel < relnums[x]){
    # set initial conditions to the end of the previous simulation and add the release
    input$init <- unlist(c(simdat[nrow(simdat),1:3],0,simdat[nrow(simdat),4],0,simdat[nrow(simdat),5:7]))
    input$init[1] <- input$init[1] + rel_size
    # simulate first generation after new release and rbind with the rest of the generations
    simdat <- rbind(simdat[1:(nrow(simdat)-1),],simfunc(input))
    rel <- rel + 1
  }
  # simulate until generation 120
  input$tendin <- 120 - nrow(simdat)
  input$init <- unlist(c(simdat[nrow(simdat),1:3],0,simdat[nrow(simdat),4],0,simdat[nrow(simdat),5:7]))
  simdat <- rbind(simdat[1:(nrow(simdat)-1),],simfunc(input))

  simdat$time <- seq(0,nrow(simdat)-1)
  simdat$letter <- LETTERS[1:4][x]
  simdat
  }
  ) %>% bind_rows()

simdat_all$letter <- factor(simdat_all$letter)
levels(simdat_all$letter) <- c("Single 2:1","2 times 2:1",
                               "5 times 2:1","5 times 1:2")
# number of generations that wild-type frequency is less than 5%
levels(simdat_all$letter) %>% sapply(function(x) simdat_all %>% 
                                       subset(simdat_all$letter==x & kkrr < 0.05) %>%
                                       nrow)

# calculate R and K allele frequencies
simdat_all$sums <- rowSums(simdat_all[,unique(c(grep("k",names(simdat_all)),grep("K",names(simdat_all))))])
simdat_all$R <- (rowSums(simdat_all[,grep("RR",names(simdat_all))]) + 
               1/2*rowSums(simdat_all[,grep("Rr",names(simdat_all))]))/ simdat_all$sums
simdat_all$K <- (rowSums(simdat_all[,grep("KK",names(simdat_all))]) + 
               1/2*rowSums(simdat_all[,grep("Kk",names(simdat_all))]))/ simdat_all$sums

# letter labels for plot
labeldf <- data.frame(time=3,frequency=1.04,letter=levels(simdat_all$letter),mylabel=LETTERS[1:4])
# plot
simdat_all %>% 
  melt(id.vars=c("time","replicate","simtype","experiment","letter"),
       measure.vars=c("R","K","kkrr"),
       variable.name="quantity",value.name="frequency") %>%
  ggplot(aes(x=time,y=frequency,size=simtype,linetype=quantity,color=quantity)) +
  geom_line(size=0.9) +
  facet_grid(letter ~ .) +
  scale_x_continuous(name="generation",limits=c(0,120),breaks = seq(0,120,20),expand = c(0, 0)) +
  scale_linetype_manual(labels=c("R"="R allele","K"="K allele","kkrr"="kkrr"),
                        values=c("R"="dashed","K"="dotted","kkrr"="dotdash")) +
  scale_color_manual(labels=c("R"="R allele","K"="K allele","kkrr"="kkrr"),
                     values=c("R"="blue","K"="red","kkrr"="black")) +
  guides(linetype=guide_legend(keywidth = 3, keyheight = 1)) +
  geom_text(data=labeldf,aes(label=mylabel,group=NULL,linetype=NULL),color="black",size=5) +
  my_theme() + #in KR_functions.R
  theme(legend.title.align = 0.5,
        legend.key = element_rect(colour = NA))
